wdl-lint/src/rules/redundant_assignment.rs

```diff
@@ -0,0 +1,120 @@
+//! A lint rule for flagging redundant `= None` assignments for
+//! optional inputs.
+
+use wdl_analysis::Diagnostics;
+use wdl_analysis::VisitReason;
+use wdl_analysis::Visitor;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::Diagnostic;
+use wdl_ast::Span;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::v1::BoundDecl;
+use wdl_ast::v1::Expr;
+use wdl_ast::v1::LiteralExpr;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the redundant assignment rule.
+const ID: &str = "RedundantAssignment";
+
+/// Create a "redundant `= None` assignment" diagnostic
+fn redundant_none(span: Span, name: &str) -> Diagnostic {
+    Diagnostic::note(format!(
+        "redundant assignment of `None` to optional input `{name}`"
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(format!("remove `= None` for input `{name}`"))
+}
+
+/// A rule that identifies redundant `= None` assignments for
+/// optional inputs.
+#[derive(Debug, Default, Clone)]
+pub struct RedundantAssignment;
+
+impl Rule for RedundantAssignment {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Flags redundant explicit assignment of `None` to optional inputs."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "The specification states that an optional input declaration (e.g., `String? foo`) is \
+         implicitly initialized to `None` if no other value or default is provided. Therefore \
+         explicitly writing `String? foo = None` is equivalent to `String? foo` but adds \
+         unnecessary verbosity. This rule encourages omitting the redundant `= None` for optional \
+         inputs.
+
+         For example, 
+        `String? my_optional_input = None` should be simplified to `String? my_optional_input`."
+    }
+
+    fn tags(&self) -> crate::TagSet {
+        TagSet::new(&[Tag::Style, Tag::Clarity])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::InputSectionNode,
+            SyntaxKind::BoundDeclNode,
+        ])
+    }
+
+    fn related_rules(&self) -> &[&'static str] {
+        &[]
+    }
+}
+
+impl Visitor for RedundantAssignment {
+    fn reset(&mut self) {
+        *self = Default::default();
+    }
+
+    fn bound_decl(
+        &mut self,
+        diagnostics: &mut Diagnostics,
+        reason: VisitReason,
+        decl: &wdl_ast::v1::BoundDecl,
+    ) {
+        if reason == VisitReason::Enter {
+            check_for_redundant_none(diagnostics, decl, &self.exceptable_nodes())
+        }
+    }
+}
+
+/// Checks for a redundant `None` assignment.
+fn check_for_redundant_none(
```

Comment:
Is there a reason for making this a function instead of just putting it all in the visitor callback?

Reply:
It's not a big deal one way or the other, but I'd argue keeping it all in the visitor callback is slightly preferable in this case. The function body can essentially be copy and pasted into the visitor callback without any changes.

Without having reviewed every case of the `check_` pattern, I think most of those probably still contain some amount of logic within the visitor callback or are called from multiple callbacks.